chain/epoch_manager/src/shard_assignment.rs

```diff
@@ -0,0 +1,274 @@
+            result[shard_id].push(cp);
+        }
+
+        let mut cp_iter = cp_iter.peekable();
+        if cp_iter.peek().is_some() {
+            // If there are still validators left to assign once the minimum is reached then
+            // we can change priorities to try and balance the total stake in each shard
+
+            // re-index shards to favour lowest stake first
+            let mut shard_index: MinHeap<(Balance, usize, ShardId)> = shard_index
+                .into_iter()
+                .map(|(validator_count, stake, shard_id)| (stake, validator_count, shard_id))
+                .collect();
+
+            for cp in cp_iter {
+                let (least_stake, validator_count, shard_id) =
+                    shard_index.pop().expect("shard_index should never be empty");
+                shard_index.push((least_stake + cp.get_stake(), validator_count + 1, shard_id));
+                result[shard_id].push(cp);
+            }
+        }
+    } else {
+        // In this case there are not enough validators to fill all shards without repeats,
+        // so we use a more complex assignment function.
+        let mut cp_iter =
+            chunk_producers.into_iter().cycle().enumerate().take(required_validator_count);
+
+        // This function assigns validators until all shards are filled. Each validator may
+        // be assigned to multiple shards. This function does not attempt to balance stakes between
+        // shards because it is expected that in production there will always be many more chunk
+        // producers than shards, so this case will never come up and thus the algorithm does not
+        // need to be tuned for this situation.
+        assign_with_possible_repeats(
+            &mut shard_index,
+            &mut result,
+            &mut cp_iter,
+            num_shards,
+            min_validators_per_shard,
+            num_chunk_producers,
+        );
+    }
+
+    Ok(result)
+}
+
+fn assign_with_possible_repeats<T: HasStake + Eq, I: Iterator<Item = (usize, T)>>(
+    shard_index: &mut MinHeap<(usize, Balance, ShardId)>,
+    result: &mut Vec<Vec<T>>,
+    cp_iter: &mut I,
+    num_shards: usize,
+    min_validators_per_shard: usize,
+    num_chunk_producers: usize,
+) {
+    let mut buffer = Vec::with_capacity(num_shards);
+
+    while shard_index.peek().unwrap().0 < min_validators_per_shard {
+        let (assignment_index, cp) = cp_iter
+            .next()
+            .expect("cp_iter should contain enough elements to minimally fill each shard");
+        let (least_validator_count, shard_stake, shard_id) =
+            shard_index.pop().expect("shard_index should never be empty");
+
+        if assignment_index < num_chunk_producers {
+            // no need to worry about duplicates yet; still on first pass through validators
+            shard_index.push((least_validator_count + 1, shard_stake + cp.get_stake(), shard_id));
```

Comment:
Between here, and two similar places above, you use inconsistent naming for the first two values you pop (you arbitrarily use or not use the word `least`).